### PR TITLE
refactor: migrate to typescript

### LINKED ISSUE #83: Migrate to Typescript
Because type-safe is better for library

### PR DESCRIPTION
**Pull request by jefrydco**
This will make breaking changes because it recreates the API from the ground up optimizing
typescript usage.

BREAKING CHANGE: One known breaking change is we can't import each of API partially because the initialization of each storage class is using a singleton.

Closes #83

### CODE REVIEW

File: src/unified-storage.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    }
+    throw new Error(`localStorage doesn't supported`)
+  }
+}
+
+export class SessionStorageFactory {
+  static storage: UnifiedStorageAbstractFactory
+
+  static create(): StorageImpl {
+    if ('sessionStorage' in window && window.sessionStorage) {
```
Comment (mazipan) [inline src/unified-storage.ts]: Adding check `window` availability, in SSR will not work.

Comment (jefrydco) [reply src/unified-storage.ts]: I've updated to check `window` availability

**Comment by mazipan** [comment]
Help resolve conflict ya @jefrydco

**Comment by jefrydco** [comment]
Resolved @mazipan

**Pull request merged**